app/scripts/modules/core/task/task.write.service.js

```diff
@@ -19,11 +20,38 @@ module.exports = angular
     }
 
     function cancelTask(application, taskId) {
-      return getEndpoint(application).one(taskId).one('cancel').put();
+      return getEndpoint(application).one(taskId).one('cancel').put().then(() =>
+        waitUntilTaskMatches(application, taskId, (task) => task.status === 'CANCELED')
+      );
+    }
+
+    function waitUntilTaskMatches(application, taskId, closure) {
+      return taskReader.getOneTaskForApplication(application, taskId).then((task) => {
+        if (closure(task)) {
+          return task;
+        }
+        return $timeout(() => waitUntilTaskMatches(application, taskId, closure), 1000);
+      });
+    }
+
+    function waitUntilTaskIsDeleted(taskId) {
+      var deferred = $q.defer();
+      Restangular.one('tasks', taskId).get().then(
+        () => {
+          $timeout(() => {
+            waitUntilTaskIsDeleted(taskId).then(() => deferred.resolve());
+          }, 1000);
+        },
+        (resp) => {
+          if (resp.status === 404) {
+            deferred.resolve();
+          }
+        });
+      return deferred.promise;
```

Comment:
Added some comments in the code and test to explain what's happening, since it's a little counterintuitive (we need to resolve when the response status is 404), and added an additional guard against other error status codes.